### ISSUE TITLE
[UX] Provide pseudo-tabs for editing node blocks from the frontend

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4925]
While logged in, people with permissions to edit content are provided with a set of "View"/"Edit"/"Revisions"/"Translate" etc. tabs when directly viewing the content:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106351880-b21acb80-62e7-11eb-81e3-82fb533290e9.png">

If a piece of content is added in a layout as a node block ("Existing content" block), there are no such tabs/links. The only easy way to edit the content is via the contextual links:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106351435-ce693900-62e4-11eb-9c8f-b221f3ab4577.png">

That assumes a) that the Contextual Links core module is enabled, and b) that the specific user role has been granted the permission to use them. If any of these assumptions are not true, then the user has to go pogo-sticking around the admin UI: navigate to the content listing (assuming they've realized that this is a node and not a custom block) -> search for the content (assuming that the content title has not been hidden) -> edit.

I propose that we allow pseudo-tabs to be rendered at the top of these node blocks - not only "Edit", but also "Revisions", "Translation" etc. (with proper permission control of course).

**Comment by indigoxela** [comment]
> I propose that we allow pseudo-tabs to be rendered at the top of these node blocks

Interesting proposal. Although I understand the intention, I'd like to note that the space for tabs might not always be available (like for nodes (snippets) in a sidebar). And if there are several nodes (snippets) on a page, that might be a visual regression, as tabs can be pretty dominant. Not sure, more feedback would be helpful.

**Comment by klonos** [comment]
> ...space for tabs might not always be available

Good catch ...in Seven, we collapse horizontal primary tabs if not enough horizontal space:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106364249-5592cd00-6336-11eb-8f42-cfcd6c0da632.png">

...and for a node:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106364265-6e02e780-6336-11eb-9107-e654384afdc4.png">

But it doesn't work properly with Basis, or other themes 😞:

<img width="50%" height="50%" src="https://user-images.githubusercontent.com/2423362/106364316-c508bc80-6336-11eb-8002-19abcaadaab7.png">

**Comment by docwilmot** [comment]
Could we just add an "Edit" link if contextual links module is disabled?

**Comment by klonos** [comment]
@docwilmot the contextual links also provide a link to direct-delete the node; and I also opened #4926 to add any other primary tab actions as contextual links. So I would prefer if we added all available contextual links, like so perhaps:

[Edit]() | [Delete]() | [Revisions]() | [Translate]()

Perhaps that way would be less prone to "narrow width distortions".

**Comment by philsward** [comment]
What about enforcing the admin label on blocks, adding admin labels to nodes (another issue discusses making admin labels more consistent across the platform), then have the "edit" tab on hover, show the list of editable blocks/nodes using the admin label as the link text?

**Comment by klonos** [comment]
> then have the "edit" tab on hover

That's more or less how contextual links work. The point is to have these action tabs/links shown without need to hover.

> The catch, is making it mobile friendly.

Exactly. That's another shortcoming of contextual links (besides that they may be disabled).

**Comment by philsward** [comment]
> That's more or less how contextual links work.

Understood, but unless you know they are there, you don't know to look for them. Unless you know how they work, you don't know to hover the region to show them.

Hovering over "edit" is far more intuitive as a "contextual link" than the contextual links themselves. Just saying.

@jenlampton mentioned in the recent weekly meetings that she would like to see contextual links overhauled as it is a poor UX inherited from Drupal 7 and Drupal 8+ has moved on to a "page builder" where it shows each region that can be manipulated when the builder is turned on. I want to say the WP admin theme handles it this way as well.

If we took the tab -> hover approach, it would actually make sense to show a "layout" tab where the various layouts that can manipulate the given path are shown.

I'm trying to look at the big picture but throwing out some ideas that might be good (medium to long term) temporary solutions.

**Comment by indigoxela** [comment]
Some more thoughts.

Regarding the situation, when contextual links are disabled... The module is enabled by default. So if it's disabled, this has probably been an admin decision (to prevent the direct on-page edit links for whatever reason) - shouldn't we respect that decision?

Regarding contextual links and touch devices (mobile or not). Yes, I also see some potential for improvement for devices without hover.

Personally, I like the unobtrusive look of contextual links - editors get a good impression, what the page looks like for visitors. If we add some links or tabs, that are always visible, I'd prefer them to be as discreet as possible.

**Comment by philsward** [comment]
@indigoxela would you be opposed to a "page builder" or "page editor" type of functionality where you click a toggle button (switch?) and it shows / hides the editable regions and their configuration options?

This would solve mobile issues I believe and also solve the problem of "as the user sees it" vs "as the editor needs it".

**Comment by indigoxela** [comment]
> @indigoxela would you be opposed to a "page builder" or "page editor" type of functionality where you click a toggle button (switch?) and it shows / hides the editable regions and their configuration options?

Oh, that seems like an interesting approach. Can you provide a mockup? (No, not opposed at all. :wink:)
